File: tools/config_manager.py
"""
config_manager.py - 配置文件管理模块

负责读取和写入 TOML 格式的配置文件（config/secrets.toml）。
包含登录凭据、Cookie 和下载选项等配置项。
"""

import os

import toml

# 配置文件的绝对路径（相对于项目根目录）
_CONFIG_FILE_PATH = os.path.join(os.getcwd(), "config", "secrets.toml")

# 首次使用时写入的默认配置
_DEFAULT_CONFIG = {
    "cookie": {
        "PHPSESSID": "",
        "jieqiUserInfo": "",
        "jieqiVisitInfo": "",
        "cf_clearance": "",
    },
    "login": {
        "username": "",
        "password": "",
    },
    "proxy": {
        "http": "",
        "https": "",
    },
    "download": {
        "full_title": "FULL",
        "default_cover_index": 0,
    },
}


class ConfigManager:
    """
    配置管理器，封装对 TOML 配置文件的读写操作。

    配置文件不存在时，自动使用默认值创建。
    所有修改操作均会立即持久化到磁盘。
    """

    def __init__(self):
        self.file_path = _CONFIG_FILE_PATH

        # 若配置文件不存在，则自动从模板复制一份（保留文件中的注释说明）
        if not os.path.exists(self.file_path):
            example_path = self.file_path + ".example"
            if os.path.exists(example_path):
                import shutil

                shutil.copy(example_path, self.file_path)

        self.config = self._read_toml()

        # 若依然无法读取，使用默认值初始化并写入
        if not self.config:
            self.config = _DEFAULT_CONFIG.copy()
            self._write_toml()

    # ------------------------------------------------------------------
    # 私有方法
    # ------------------------------------------------------------------

    def _read_toml(self) -> dict:
        """
        读取 TOML 配置文件。

        :return: 配置字典；文件不存在时返回空字典
        """
        if os.path.exists(self.file_path):
            with open(self.file_path, "r", encoding="utf-8") as f:
                return toml.load(f)
        return {}

    def _write_toml(self) -> None:
        """将当前内存中的配置持久化到 TOML 文件。"""
        with open(self.file_path, "w", encoding="utf-8") as f:
            toml.dump(self.config, f)

    # ------------------------------------------------------------------
    # 公开方法
    # ------------------------------------------------------------------

    def get(self, section: str, key: str = None):
        """
        获取配置项的值。

        :param section: 配置节名称，如 "login"、"cookie"
        :param key: 节内键名；若为 None，则返回整个节
        :return: 对应的值；section 或 key 不存在时返回 None
        """
        if section not in self.config:
            return None
        if key is not None:
            return self.config[section].get(key)
        return self.config[section]

    def set(self, section: str, key: str, value) -> None:
        """
        设置配置项的值，并立即写入文件。

        :param section: 配置节名称
        :param key: 节内键名
        :param value: 要写入的值
        """
        if section not in self.config:
            self.config[section] = {}
        self.config[section][key] = value
        self._write_toml()

    def delete(self, section: str, key: str) -> None:
        """
        删除指定配置项，并立即写入文件。

        :param section: 配置节名称
        :param key: 节内键名
        """
        if section in self.config and key in self.config[section]:
            del self.config[section][key]
            self._write_toml()

    def get_all(self) -> dict:
        """返回完整的配置字典（只读引用）。"""
        return self.config

File: src/main/config-manager.ts
import { readFileSync, writeFileSync, existsSync, mkdirSync, renameSync, chmodSync } from 'fs'
import { join } from 'path'
import { createCipheriv, createDecipheriv, createHash, randomBytes } from 'crypto'
import { parse, stringify } from 'smol-toml'
import type { Wenku8Config } from './types'

function getConfigDir(): string {
  const webDataDir = process.env.WEB_DATA_DIR
  if (webDataDir) {
    return join(webDataDir, 'config')
  }

  try {
    const { app } = require('electron') as typeof import('electron')
    if (!app.isPackaged) {
      return join(process.cwd(), '.dev-user-data', 'config')
    }
    return join(app.getPath('userData'), 'config')
  } catch {
    return join(process.cwd(), '.dev-user-data', 'config')
  }
}

const CONFIG_DIR = getConfigDir()
const IS_WEB_RUNTIME = Boolean(process.env.WEB_DATA_DIR)
const CONFIG_PATH = join(CONFIG_DIR, IS_WEB_RUNTIME ? 'secrets.enc' : 'secrets.toml')

const DEFAULT_CONFIG: Wenku8Config = {
  cookie: {
    PHPSESSID: '',
    jieqiUserInfo: '',
    jieqiVisitInfo: '',
    cf_clearance: '',
    userAgent: '',
  },
  login: {
    username: '',
    password: '',
  },
  download: {
    full_title: 'FULL',
    default_cover_index: 0,
    download_path: '',
  },
  proxy: {
    enabled: false,
    url: '',
  },
}

const PROXY_PROTOCOLS = new Set(['http:', 'https:', 'socks5:', 'socks5h:'])

export function normalizeProxyUrl(value: string): string {
  const input = value.trim()
  if (!input) return ''
  if (input.length > 2048 || /[\u0000-\u001f\u007f]/.test(input)) {
    throw new Error('代理地址无效')
  }

  let url: URL
  try {
    url = new URL(input)
  } catch {
    throw new Error('代理地址无效')
  }
  if (
    !PROXY_PROTOCOLS.has(url.protocol) ||
    !url.hostname ||
    (url.pathname && url.pathname !== '/') ||
    url.search ||
    url.hash
  ) {
    throw new Error('仅支持 HTTP、HTTPS、SOCKS5 代理地址')
  }
  return url.toString()
}

export function redactProxyUrl(value: string): { url: string; hasCredentials: boolean } {
  if (!value) return { url: '', hasCredentials: false }
  const url = new URL(value)
  const hasCredentials = Boolean(url.username || url.password)
  url.username = ''
  url.password = ''
  return { url: url.toString(), hasCredentials }
}

class ConfigManager {
  private config: Wenku8Config

  constructor() {
    mkdirSync(CONFIG_DIR, { recursive: true })

    if (!existsSync(CONFIG_PATH)) {
      this.config = structuredClone(DEFAULT_CONFIG)
      this.writeConfig()
    } else {
      this.config = this.readConfig()
      if (!this.config || Object.keys(this.config).length === 0) {
        this.config = structuredClone(DEFAULT_CONFIG)
        this.writeConfig()
      }
    }
    this.config = {
      ...structuredClone(DEFAULT_CONFIG),
      ...this.config,
      cookie: { ...DEFAULT_CONFIG.cookie, ...this.config.cookie },
      login: { ...DEFAULT_CONFIG.login, ...this.config.login },
      download: { ...DEFAULT_CONFIG.download, ...this.config.download },
      proxy: { ...DEFAULT_CONFIG.proxy, ...this.config.proxy },
    }
  }

  private getEncryptionKey(): Buffer {
    const secret = process.env.APP_SECRET
    if (!secret || secret.length < 32) {
      throw new Error('网页端需要设置至少 32 个字符的 APP_SECRET')
    }
    return createHash('sha256').update(secret).digest()
  }

  private readConfig(): Wenku8Config {
    try {
      const raw = readFileSync(CONFIG_PATH, 'utf-8')
      if (IS_WEB_RUNTIME) {
        const payload = JSON.parse(raw) as { iv: string; tag: string; data: string }
        const decipher = createDecipheriv(
          'aes-256-gcm',
          this.getEncryptionKey(),
          Buffer.from(payload.iv, 'base64'),
        )
        decipher.setAuthTag(Buffer.from(payload.tag, 'base64'))
        const decrypted = Buffer.concat([
          decipher.update(Buffer.from(payload.data, 'base64')),
          decipher.final(),
        ]).toString('utf-8')
        return parse(decrypted) as unknown as Wenku8Config
      }
      return parse(raw) as unknown as Wenku8Config
    } catch (error) {
      if (IS_WEB_RUNTIME) throw error
      return {} as Wenku8Config
    }
  }

  private writeConfig(): void {
    const toml = stringify(this.config as unknown as Record<string, unknown>)
    const tempPath = `${CONFIG_PATH}.tmp`
    let content = toml

    if (IS_WEB_RUNTIME) {
      const iv = randomBytes(12)
      const cipher = createCipheriv('aes-256-gcm', this.getEncryptionKey(), iv)
      const encrypted = Buffer.concat([cipher.update(toml, 'utf-8'), cipher.final()])
      content = JSON.stringify({
        iv: iv.toString('base64'),
        tag: cipher.getAuthTag().toString('base64'),
        data: encrypted.toString('base64'),
      })
    }

    writeFileSync(tempPath, content, { encoding: 'utf-8', mode: 0o600 })
    renameSync(tempPath, CONFIG_PATH)
    if (IS_WEB_RUNTIME) chmodSync(CONFIG_PATH, 0o600)
  }

  get(section: string, key?: string): unknown {
    const sec = (this.config as unknown as Record<string, Record<string, unknown>>)[section]
    if (!sec) return null
    if (key !== undefined) return sec[key] ?? null
    return sec
  }

  set(section: string, key: string, value: unknown): void {
    const cfg = this.config as unknown as Record<string, Record<string, unknown>>
    if (!cfg[section]) cfg[section] = {}
    if (section === 'proxy' && key === 'url') {
      const proxyUrl = normalizeProxyUrl(String(value ?? ''))
      cfg.proxy.url = proxyUrl
      if (!proxyUrl) cfg.proxy.enabled = false
    } else if (section === 'proxy' && key === 'enabled') {
      const enabled = value === true || value === 'true'
      if (enabled && !cfg.proxy.url) {
        throw new Error('请先填写代理地址')
      }
      cfg.proxy.enabled = enabled
    } else if (section === 'download' && key === 'default_cover_index') {
      const index = Number(value)
      if (!Number.isInteger(index) || index < 0) {
        throw new Error('封面图片索引必须为非负整数')
      }
      cfg[section][key] = index
    } else {
      cfg[section][key] = value
    }
    this.writeConfig()
  }

  delete(section: string, key: string): void {
    const cfg = this.config as unknown as Record<string, Record<string, unknown>>
    if (cfg[section] && key in cfg[section]) {
      delete cfg[section][key]
      this.writeConfig()
    }
  }

  getAll(): Wenku8Config {
    return this.config
  }
}

export const config = new ConfigManager()
